### PR TITLE
Add UInt256, serialization methods and CLR updates

### PR DESCRIPTION
**Pull request by quantumagi**

### CODE REVIEW

**Comment by fassadlr** [comment]
@quantumagi the solution doesnt build :)

**Comment by quantumagi** [comment]
@fassadlr , yes we're going to have to re-build some of the smart contract NuGet packages before this works.

**Comment by YakupIpek** [comment]
@quantumagi Is this ready to review ?

**Comment by quantumagi** [comment]
@YakupIpek , yes.

**Review by YakupIpek — CHANGES_REQUESTED** [review]

File: src/Stratis.SmartContracts.CLR/Serialization/Serializer.cs
Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR/Serialization/Serializer.cs:83]: See https://github.com/stratisproject/Stratis.SmartContracts/pull/10.

File: src/Stratis.SmartContracts.CLR/Serialization/Serializer.cs
Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR/Serialization/Serializer.cs:89]: The code was a temporary workaround to make serialization work despite missing interface methods.

File: src/Stratis.SmartContracts.CLR/Serialization/ContractPrimitiveSerializer.cs
Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR/Serialization/ContractPrimitiveSerializer.cs]: Done

**Comment by quantumagi** [comment]
@YakupIpek , this should be good to merge once the SmartContract package references have been updated to 1.4.1-alpha and the workaround code removed.

**Comment by quantumagi** [comment]
Should be good to go once approved.

File: src/Stratis.SmartContracts.CLR/Serialization/Serializer.cs
Comment (YakupIpek) [reply src/Stratis.SmartContracts.CLR/Serialization/Serializer.cs]: I understand it now

**Comment by quantumagi** [comment]
Fixed.

File: src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs
Comment (YakupIpek) [inline src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:21]: As far as i understand these type of changes will cause conflicts between new and old nodes

Comment (YakupIpek) [reply src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:21]: If it is just because of `Stratis.SmartContracts.Standards` so we can include that older interface IStandardToken and rename the new one as IStandard256

Comment (YakupIpek) [reply src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:21]: I think the conflicts will be resolved if we can still provide same IStandardToken interface and implementation as old and add new ones with suffix 256. 

But still this is going to be issue between new and old nodes whenever new contract submited by uint256 and uint128 types.

File: src/Stratis.SmartContracts.CLR/Serialization/ContractPrimitiveSerializer.cs
Comment (YakupIpek) [inline src/Stratis.SmartContracts.CLR/Serialization/ContractPrimitiveSerializer.cs:147]: I recommends switch expression syntax 

```
            return Type.GetTypeCode(type) switch
            {
                TypeCode.Byte => stream[0],
                TypeCode.Char => ToChar(stream),
                TypeCode.Boolean => ToBool(stream),
                TypeCode.Int32 => ToInt32(stream),
                TypeCode.Int64 => ToInt64(stream),
                TypeCode.String => ToString(stream),
                TypeCode.UInt32 => ToUInt32(stream),
                TypeCode.UInt64 => ToUInt64(stream),
                _ when type == typeof(UInt128) => ToUInt128(stream),
                _ when type == typeof(uint256) => ToUInt256(stream),
                _ when type == typeof(Address) => ToAddress(stream),
                _ when type.IsArray => DeserializeArray(type.GetElementType(), stream),
                _ when type.IsValueType => DeserializeStruct(type, stream),
                _ => throw new ContractPrimitiveSerializationException(string.Format("{0} is not supported.", type.Name)),
            };
```

File: src/Stratis.SmartContracts.CLR/Serialization/ContractPrimitiveSerializer.cs
Comment (YakupIpek) [inline src/Stratis.SmartContracts.CLR/Serialization/ContractPrimitiveSerializer.cs:158]: These lines seems removed wrongly but anyway i added it in my example

File: src/Stratis.SmartContracts.CLR/Serialization/MethodParameterStringSerializer.cs
Comment (YakupIpek) [inline src/Stratis.SmartContracts.CLR/Serialization/MethodParameterStringSerializer.cs:111]:
```
  return o switch
  {
      bool _ => MethodParameterDataType.Bool,
      byte _ => MethodParameterDataType.Byte,
      byte[] _ => MethodParameterDataType.ByteArray,
      char _ => MethodParameterDataType.Char,
      string _ => MethodParameterDataType.String,
      uint _ => MethodParameterDataType.UInt,
      ulong _ => MethodParameterDataType.ULong,
      Address _ => MethodParameterDataType.Address,
      long _ => MethodParameterDataType.Long,
      int _ => MethodParameterDataType.Int,
      UInt128 _ => MethodParameterDataType.UInt128,
      UInt256 _ => MethodParameterDataType.UInt256,
      _ => throw new MethodParameterStringSerializerException(string.Format("{0} is not supported.", o.GetType().Name))
  };
```

File: src/Stratis.SmartContracts.CLR/Serialization/Prefix.cs
Comment (YakupIpek) [inline src/Stratis.SmartContracts.CLR/Serialization/Prefix.cs:49]: return ((MethodParameterDataType)b) switch
            {
                MethodParameterDataType.Address => typeof(Address),
                MethodParameterDataType.Bool => typeof(bool),
                MethodParameterDataType.Byte => typeof(byte),
                MethodParameterDataType.Char => typeof(char),
                MethodParameterDataType.String => typeof(string),
                MethodParameterDataType.Int => typeof(int),
                MethodParameterDataType.UInt => typeof(uint),
                MethodParameterDataType.Long => typeof(long),
                MethodParameterDataType.ULong => typeof(ulong),
                MethodParameterDataType.ByteArray => typeof(byte[]),
                MethodParameterDataType.UInt128 => typeof(UInt128),
                MethodParameterDataType.UInt256 => typeof(UInt256),
                _ => throw new ArgumentOutOfRangeException(nameof(b), b, "Unsupported type"),
            };

File: src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs
Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:21]: Au contraire! The change is meant to prevent old smart contracts from running with the new incompatible version of the `IStandardToken`  interface. 

The only "issue", which is unrelated to this change, is that old nodes will not understand contracts based off the new datatypes.

Could you perhaps elaborate on what you mean?

File: src/Stratis.SmartContracts.CLR/Serialization/ContractPrimitiveSerializer.cs
Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR/Serialization/ContractPrimitiveSerializer.cs:147]: This is unrelated to this PR so normally I would not make this change. Let me add it anyway since you did invest some time.

File: src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs
Comment (YakupIpek) [inline src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:23]: Possible to append version suffix to name, 64 bit seems sounds like compilation of dll for 32 bit vs 64 bit systems, WDYT?

**Comment by quantumagi** [comment]
WARNING: This PR has to be carefully managed to avoid a hard fork. Merge after a plan is in place.

Note that PR #392 may smoothen the upgrade path if released well ahead of this PR.

File: src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs
Comment (YakupIpek) [inline src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:43]: Downloading assembly packages from a central resource is not safe and it is against single point of failure and also it is a security issue.

Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:43]: Then I suggest that we simply pop the message and have the user do this manually. WDYT?

Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:44]: I've added some code to limit what can be done in the standards package - similar to what we do for the smart contracts.

Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR/Loader/ContractAssemblyLoader.cs:44]: Since there is an approval process involved with smart contracts another approach would be to release the upgraded standards package and simply not approve new contracts based on it until a sufficient number of node have upgraded. This way the nodes don't have to pull any package on demand and can simply use the appropriate package from the LegacyStandardsDLLs folder.

**Comment by quantumagi** [comment]
Versions should perhaps be bumped to 1.0.6.4 as the last step before merging this.

File: src/Stratis.Bitcoin.Features.SmartContracts.Tests/Stratis.Bitcoin.Features.SmartContracts.Tests.csproj
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.SmartContracts.Tests/Stratis.Bitcoin.Features.SmartContracts.Tests.csproj:21]: These version needs to be inline with FN (1.0.6.4)

File: src/Stratis.Bitcoin.Features.SmartContracts/Stratis.Bitcoin.Features.SmartContracts.csproj
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.SmartContracts/Stratis.Bitcoin.Features.SmartContracts.csproj:17]: These version needs to be inline with FN (1.0.6.4)

File: src/Stratis.SmartContracts.CLR.Validation/Stratis.SmartContracts.CLR.Validation.csproj
Comment (fassadlr) [inline src/Stratis.SmartContracts.CLR.Validation/Stratis.SmartContracts.CLR.Validation.csproj:12]: These version needs to be inline with FN (1.0.6.4)

**Review by fassadlr — COMMENTED** [review]
Need to fix versions so long.

File: src/Stratis.Bitcoin.Features.SmartContracts.Tests/Stratis.Bitcoin.Features.SmartContracts.Tests.csproj
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.SmartContracts.Tests/Stratis.Bitcoin.Features.SmartContracts.Tests.csproj:21]: Previously it seems the SCs had their own version sequence. Even Stratis.SmartContracts and Stratis.SmartContracts.Standads had different versions. I guess the advantage was the ability to release an update of just one module and not worry about the others being different. Are you sure we should try and keep versions in step?

File: src/Stratis.Bitcoin.Features.SmartContracts/Stratis.Bitcoin.Features.SmartContracts.csproj
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.SmartContracts/Stratis.Bitcoin.Features.SmartContracts.csproj:17]: See above.

File: src/Stratis.SmartContracts.CLR.Validation/Stratis.SmartContracts.CLR.Validation.csproj
Comment (quantumagi) [reply src/Stratis.SmartContracts.CLR.Validation/Stratis.SmartContracts.CLR.Validation.csproj:12]: See above.

**Comment by YakupIpek** [comment]
@quantumagi LGTM

**Review by YakupIpek — APPROVED** [review]

**Pull request merged**
